core/src/tpu_entry_notifier.rs

```diff
@@ -0,0 +1,85 @@
+use {
+    crossbeam_channel::{Receiver, RecvTimeoutError, Sender},
+    solana_entry::entry::EntrySummary,
+    solana_ledger::entry_notifier_service::{EntryNotification, EntryNotifierSender},
+    solana_poh::poh_recorder::{WorkingBankEntry, WorkingBankEntryWithIndex},
+    std::{
+        sync::{
+            atomic::{AtomicBool, Ordering},
+            Arc,
+        },
+        thread::{self, Builder, JoinHandle},
+        time::Duration,
+    },
+};
+
+pub(crate) struct TpuEntryNotifier {
+    thread_hdl: JoinHandle<()>,
+}
+
+impl TpuEntryNotifier {
+    pub(crate) fn new(
+        receiver: Receiver<WorkingBankEntryWithIndex>,
+        entry_notification_sender: Option<EntryNotifierSender>,
+        broadcast_entry_sender: Sender<WorkingBankEntry>,
+        exit: &Arc<AtomicBool>,
+    ) -> Self {
+        let exit = exit.clone();
+        let thread_hdl = Builder::new()
+            .name("solTpuEntry".to_string())
+            .spawn(move || loop {
+                if exit.load(Ordering::Relaxed) {
+                    break;
+                }
+
+                if let Err(RecvTimeoutError::Disconnected) = Self::send_entry_notification(
+                    &receiver,
+                    &entry_notification_sender,
+                    &broadcast_entry_sender,
+                ) {
+                    break;
+                }
+            })
+            .unwrap();
+        Self { thread_hdl }
+    }
+
+    pub(crate) fn send_entry_notification(
+        receiver: &Receiver<WorkingBankEntryWithIndex>,
+        entry_notification_sender: &Option<EntryNotifierSender>,
+        broadcast_entry_sender: &Sender<WorkingBankEntry>,
+    ) -> Result<(), RecvTimeoutError> {
+        let (bank, (entry, tick_height), index) = receiver.recv_timeout(Duration::from_secs(1))?;
+        let slot = bank.slot();
+        let index = index.unwrap_or(0);
+
+        if let Some(entry_notification_sender) = entry_notification_sender {
+            let entry_summary = EntrySummary {
+                num_hashes: entry.num_hashes,
+                hash: entry.hash,
+                num_transactions: entry.transactions.len() as u64,
+            };
+            if let Err(err) = entry_notification_sender.send(EntryNotification {
+                slot,
+                index,
+                entry: entry_summary,
+            }) {
+                warn!(
+                    "Failed to send slot {slot:?} entry {index:?} from Tpu to EntryNotifierService, error {err:?}",
+                );
+            }
+        }
+
+        // TODO: in PohRecorder, we panic if the send to BroadcastStage fails. Should we do the same here?
```

Comment:
Thoughts on this question?
Also, would you want to see this send in a separate thread?

Reply:
> do we panic in pohrecorder? PohService waits for an exit signal https://github.com/apfitzge/solana/blob/abbf96f17aca9647766559db5b5de714726da033/poh/src/poh_service.rs#L371, and poh_recorder doesn't panic on record either

This is the panic I was referring to: https://github.com/solana-labs/solana/blob/37ebb709e6f2f0c594d6187c0deccda17c742916/poh/src/poh_recorder.rs#L186

It's a bit of a chore to track all the results around, so I might be misreading. But I think that if the BroadcastStage send fails with an Err, [this](https://github.com/solana-labs/solana/blob/37ebb709e6f2f0c594d6187c0deccda17c742916/poh/src/poh_service.rs#L258) returns an Err, which gets returned to that method with the panic I linked above [here](https://github.com/solana-labs/solana/blob/37ebb709e6f2f0c594d6187c0deccda17c742916/poh/src/poh_recorder.rs#L232). Is that right?